src/commands/redis_cmd.cc

```diff
@@ -2554,152 +2556,135 @@ class CommandZPopMax : public CommandZPop {
-    rocksdb::Status s = zset_db.Range(args_[1], start_, stop_, flags, &member_scores);
+    rocksdb::Status s;
+    int size = 0;
+    if (by_flag_ == "BYLEX") {
+      s = zset_db.RangeByLex(args_[1], *static_cast<ZRangeLexSpec *>(spec_.get()), &member_scores, &size);
+    } else if (by_flag_ == "BYSCORE") {
+      s = zset_db.RangeByScore(args_[1], *static_cast<ZRangeSpec *>(spec_.get()), &member_scores, &size);
+    } else if (by_flag_ == "BYINDEX") {
+      s = zset_db.RangeByIndex(args_[1], *static_cast<ZRangeIndexSpec *>(spec_.get()), &member_scores);
+    } else {
+      assert(false);
+    }
     if (!s.ok()) {
       return Status(Status::RedisExecErr, s.ToString());
     }
+    packageOutput(member_scores, output);
+    return Status::OK();
+  }
+
+ private:
+  void packageOutput(const std::vector<MemberScore> &member_scores, std::string *output) {
     if (!with_scores_) {
-      output->append(Redis::MultiLen(member_scores.size()));
+      output->append(Redis::MultiLen(static_cast<int64_t>(member_scores.size())));
     } else {
-      output->append(Redis::MultiLen(member_scores.size() * 2));
+      output->append(Redis::MultiLen(static_cast<int64_t>(member_scores.size() * 2)));
     }
     for (const auto &ms : member_scores) {
       output->append(Redis::BulkString(ms.member));
       if (with_scores_) output->append(Redis::BulkString(Util::Float2String(ms.score)));
     }
-    return Status::OK();
   }
-
- private:
-  int start_ = 0;
-  int stop_ = 0;
+  std::string_view by_flag_ = "";
   bool reversed_;
   bool with_scores_ = false;
+  std::shared_ptr<ZrangeCommon> spec_;
 };
 
 class CommandZRevRange : public CommandZRange {
  public:
   CommandZRevRange() : CommandZRange(true) {}
 };
 
-class CommandZRangeByLex : public Commander {
+class CommandZRangeByLex : public CommandZRange {
  public:
-  explicit CommandZRangeByLex(bool reversed = false) { spec_.reversed = reversed; }
+  explicit CommandZRangeByLex(bool reversed = false) : reversed_(reversed) {}
 
   Status Parse(const std::vector<std::string> &args) override {
-    Status s;
-    if (spec_.reversed) {
-      s = Redis::ZSet::ParseRangeLexSpec(args[3], args[2], &spec_);
-    } else {
-      s = Redis::ZSet::ParseRangeLexSpec(args[2], args[3], &spec_);
-    }
-    if (!s.IsOK()) {
-      return Status(Status::RedisParseErr, s.Msg());
-    }
-    if (args.size() == 7 && Util::ToLower(args[4]) == "limit") {
-      auto parse_offset = ParseInt<int>(args[5], 10);
-      auto parse_count = ParseInt<int>(args[6], 10);
-      if (!parse_offset || !parse_count) {
-        return Status(Status::RedisParseErr, errValueNotInteger);
-      }
-      spec_.offset = *parse_offset;
-      spec_.count = *parse_count;
-    }
-    return Commander::Parse(args);
+    const_cast<std::vector<std::string> &>(args).emplace_back("BYLEX");
+    if (reversed_) const_cast<std::vector<std::string> &>(args).emplace_back("REV");
+    return CommandZRange::Parse(args);
```

Comment:
I think you can find a better way to organize these code here.

To be honest, it is ugly to use `const_cast` (to break the original semantics) and takes more runtime overhead (`zrangebylex` can be more efficient to parse than `zrange`). 

And more important is, it is not the correct way to parse `zrangebylex`. For example, it will allow some commands like `zrangebylex key x y BYLEX`.

Reply:
It looks like I need to parse it all over again for these instructions. But I think using `const_cast` essentially splits the `zrangebylex` command into `zrange byindex`, which should make sense.
 I wanted to make these commands use the same code as zrange's option as much as possible, not only to facilitate debugging, but also to reflect the fact that the commands have been deprecated.